packages/headless/src/features/facets/facet-set/facet-set-analytics-actions-utils.ts

```diff
@@ -21,3 +43,158 @@ export type FacetSelectionChangeMetadata = {
+  payload: FacetSelectionChangeMetadata,
+  state: SectionNeededForFacetMetadata
+) {
+  const {facetId, facetValue} = payload;
+  const base = buildFacetBaseMetadata(facetId, state);
+
+  return {...base, facetValue};
+}
+
+export function getStateNeededForFacetMetadata(
+  s: Partial<SearchAppState>
+): SectionNeededForFacetMetadata {
+  return {
+    facetSet: s.facetSet || getFacetSetInitialState(),
+    categoryFacetSet: s.categoryFacetSet || getCategoryFacetSetInitialState(),
+    dateFacetSet: s.dateFacetSet || getDateFacetSetInitialState(),
+    numericFacetSet: s.numericFacetSet || getNumericFacetSetInitialState(),
+    search: s.search || getSearchInitialState(),
+  };
+}
+
+export const buildFacetStateMetadata = (
+  state: SectionNeededForFacetMetadata
+) => {
+  const facetState: FacetStateMetadata[] = [];
+
+  state.search.response.facets.forEach((facetResponse, facetPosition) => {
+    (facetResponse.values as Array<AnyFacetValue>).forEach(
+      (facetValue, facetValuePositition) => {
+        if (facetValue.state === 'selected') {
+          const facetType = getFacetType(state, facetResponse.facetId);
+
+          const facetResponseAnalytics = mapFacetResponseToAnalytics(
+            facetResponse,
+            facetPosition + 1
+          );
+
+          const facetValueAnalytics = mapFacetValueToAnalytics(
+            facetValue,
+            facetValuePositition + 1,
+            facetType
+          );
+
+          const facetDisplayValueAnalytics =
+            facetType === 'hierarchical' || facetType === 'specific'
+              ? mapFacetDisplayValueToAnalytics(
+                  facetValue as CategoryFacetValue | FacetValue
+                )
+              : mapRangeDisplayFacetValueToAnalytics(
+                  facetValue as NumericFacetValue | DateFacetValue
+                );
+
+          facetState.push({
+            ...facetResponseAnalytics,
+            ...facetValueAnalytics,
+            ...facetDisplayValueAnalytics,
+          });
+        }
+      }
+    );
+  });
+
+  return facetState;
+};
+
+const mapFacetValueToAnalytics = (
+  facetValue: AnyFacetValue,
+  valuePosition: number,
+  facetType: AvailableType
+) => {
+  return {
+    state: facetValue.state,
+    valuePosition,
+    facetType,
+  };
+};
+
+const mapRangeDisplayFacetValueToAnalytics = (
+  facetValue: DateFacetValue | NumericFacetValue
+) => {
+  return {
+    displayValue: `${facetValue.start}..${facetValue.end}`,
+    value: `${facetValue.start}..${facetValue.end}`,
+    start: facetValue.start,
+    end: facetValue.end,
+    endInclusive: facetValue.endInclusive,
+  };
+};
+
+const mapFacetDisplayValueToAnalytics = (
+  facetValue: CategoryFacetValue | FacetValue
+) => {
+  return {
+    displayValue: facetValue.value,
+    value: facetValue.value,
+  };
+};
+
+const mapFacetResponseToAnalytics = (
+  response: AnyFacetResponse,
+  facetPosition: number
+) => {
+  return {
+    title: response.facetId,
```

Comment:
If we use the `facetId` as the title, should we also adjust [this line here](https://github.com/coveo/ui-kit/pull/346/files?file-filters%5B%5D=.ts#diff-d727197efe0057eac84b9b6dc7b80c3b7fdf1d0ce8ae184432773940e3307177R58) to do the same?

Marie said they like having the field inside the id for debugging.
With the new id system, the field is used in the id.

The only case where it may not is if a user manually specifies their own id.